File: android/src/main/java/org/ligi/gobandroid_hd/logic/cell_gatherer/CellGatherer.java
package org.ligi.gobandroid_hd.logic.cell_gatherer;

import org.ligi.gobandroid_hd.logic.StatefulGoBoard;
import org.ligi.gobandroid_hd.logic.StatelessBoardCell;

import java.util.HashSet;
import java.util.Set;

public abstract class CellGatherer extends HashSet<StatelessBoardCell> {

    protected final Set<StatelessBoardCell> processed;
    protected final StatelessBoardCell root;
    protected final StatefulGoBoard board;

    public CellGatherer(final StatefulGoBoard board, final StatelessBoardCell root) {
        this.board = board;
        this.root = root;
        processed = new HashSet<>();
        pushWithCheck(root);
    }


    abstract protected void pushWithCheck(StatelessBoardCell cell);

    protected void pushSurroundingWithCheck(StatelessBoardCell cell) {
        for (StatelessBoardCell boardCell : cell.getNeighbors()) {
            pushWithCheck(boardCell);
        }
    }

    public Set<StatelessBoardCell> getProcessed() {
        return processed;
    }

}

File: android/src/main/java/org/ligi/gobandroid_hd/logic/cell_gatherer/CellGatherer.kt
package org.ligi.gobandroid_hd.logic.cell_gatherer

import org.ligi.gobandroid_hd.logic.StatefulGoBoard
import org.ligi.gobandroid_hd.logic.StatelessBoardCell
import java.util.*

abstract class CellGatherer(protected val board: StatefulGoBoard, protected val root: StatelessBoardCell) {

    val gatheredCells = HashSet<StatelessBoardCell>()
    val processed = HashSet<StatelessBoardCell>()

    init {
        process(root)
    }

    protected fun process(cell: StatelessBoardCell) {

        var toProcess = hashSetOf(cell)

        while (!toProcess.isEmpty()) {
            toProcess = processAndGetFollowup(toProcess)
        }

    }

    open fun processAndGetFollowup(toProcess: Set<StatelessBoardCell>): HashSet<StatelessBoardCell> {
        val followUp = HashSet<StatelessBoardCell>()
        toProcess.forEach {
            if (processed.add(it) && board.areCellsEqual(it, root)) {
                gatheredCells.add(it)
                followUp.addAll(it.neighbors)
            }
        }
        return followUp
    }



}
